### PR TITLE
Synchronize gradients in manual optimization with DDPStrategy(static_graph=True)

### LINKED ISSUE #18086: Gradients seem not synchronized with manual optimization and DDPStrategy with `static_graph=True`
### Bug description

Gradients seem not synchronized with manual optimization and DDPStrategy with `static_graph=True`.

### What version are you seeing the problem on?

v2.0.5

### How to reproduce the bug

Create `main.py`, and run `python main.py fit` with two GPUs:

```python
# main.py
from lightning.pytorch.cli import LightningCLI
from lightning.pytorch.demos.boring_classes import BoringModel
from lightning.pytorch.strategies import DDPStrategy

class MyModel(BoringModel):
    def __init__(self):
        super().__init__()
        self.automatic_optimization = False

    def training_step(self, batch, batch_idx: int):
        optimizer = self.optimizers()
        optimizer.zero_grad()
        out = super().training_step(batch, batch_idx)
        loss = out['loss']
        self.manual_backward(loss)
        optimizer.step()

    def on_train_batch_end(self, *args, **kwargs):
        print(self.layer.bias.grad, f'[rank {self.global_rank}] grad')
        print(self.layer.bias, f'[rank {self.global_rank}]')

def main():
    LightningCLI(
        MyModel,
        save_config_kwargs={'overwrite': True},
        trainer_defaults={
            'strategy': DDPStrategy(static_graph=True),
            'max_steps': 1,
            'enable_progress_bar': False,
        },
        seed_everything_default=42,
    )

if __name__ == '__main__':
    main()
```

### Error messages and logs

The outputs of the script above are as follows, the gradients are not synchronized.:
```
tensor([-1.8170, -1.3621], device='cuda:0') [rank 0] grad
tensor([-1.1449, -2.1265], device='cuda:1') [rank 1] grad
Parameter containing:
tensor([0.2359, 0.0994], device='cuda:0', requires_grad=True) [rank 0]
Parameter containing:
tensor([0.1687, 0.1758], device='cuda:1', requires_grad=True) [rank 1]
```
When setting `static_graph=False` or using automatic optimization, the outputs are as follows, the gradients are synchronized:
```
tensor([-1.4809, -1.7443], device='cuda:0') [rank 0] grad
tensor([-1.4809, -1.7443], device='cuda:1') [rank 1] grad
Parameter containing:
tensor([0.2023, 0.1376], device='cuda:0', requires_grad=True) [rank 0]
Parameter containing:
tensor([0.2023, 0.1376], device='cuda:1', requires_grad=True) [rank 1]
```

### Environment

<details>
  <summary>Current environment</summary>

```
#- PyTorch Lightning Version (e.g., 1.5.0): 2.0.5
#- PyTorch Version (e.g., 2.0): 2.0.1
#- Python version (e.g., 3.9): 3.11.4
#- OS (e.g., Linux): Linux
#- How you installed Lightning(`conda`, `pip`, source): pip
```

</details>

### More info

_No response_

cc @borda @justusschock @lantiga

### PR DESCRIPTION
**Pull request by Sohaib-Ahmed21**
## What does this PR do?

Fixes gradient synchronization when using **manual optimization** with `DDPStrategy(static_graph=True)`.

- Calls `reducer._delay_all_reduce()` in `post_backward` to ensure gradients are properly reduced in the first iteration.
- Adds a regression test (`test_ddp_gradients_synced`) that checks gradient synchronization across:
  - `automatic_optimization=True` / `False`
  - `static_graph=True` / `False`

Fixes #18086 

<details>
  <summary><b>Before submitting</b></summary>

- Was this **discussed/agreed** via a GitHub issue? (not for typos and docs)
Yes
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- Did you write any **new necessary tests**? (not for typos and docs)
yes
- [x] Did you verify new and **existing tests pass** locally with your changes?
- Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

</details>

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

<details>
  <summary>Reviewer checklist</summary>

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

</details>

----
📚 Documentation preview 📚: https://pytorch-lightning--21251.org.readthedocs.build/en/21251/

### CODE REVIEW

**Comment by Sohaib-Ahmed21** [comment]
Failures seem unrelated to the PR, can you please review the PR @Borda @SkafteNicki , thanks!

**Comment by SkafteNicki** [comment]
> Failures seem unrelated to the PR, can you please review the PR @Borda @SkafteNicki , thanks!

@Sohaib-Ahmed21 yes CI is down at the moment so nothing to do at the moment

**Comment by Sohaib-Ahmed21** [comment]
@Borda @SkafteNicki all tests are passing now. Can you please review this PR, thanks!

**Comment by Sohaib-Ahmed21** [comment]
@SkafteNicki  @Borda  @lantiga , gentle ping on this PR. I believe it's ready for review when you have a moment. Thanks!

**Comment by Borda** [comment]
> @SkafteNicki @Borda @lantiga , gentle ping on this PR. I believe it's ready for review when you have a moment. Thanks!

Apology, I am not a code-owner anymore...
